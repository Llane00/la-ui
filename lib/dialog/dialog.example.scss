.la-ui-layout {
  .dialog-example-ct {
    margin-bottom: 2em;

    h1 {
      margin: 0 0 .8em;
      font-size: 1.1em;
    }

    button {
      margin: 0 2em 0 0;
      padding: .2em .5em;
    }
  }
}