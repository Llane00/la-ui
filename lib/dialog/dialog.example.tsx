import React, { useState } from 'react';
import Dialog, { alert, confirm, modal } from './dialog';
import './dialog.example.scss';
import Button from '../button/button';

export default function () {
  const [x, setX] = useState(false); //for 例子1
  const [y, setY] = useState(false); //for 例子2
  const openModal = () => {
    const close = modal(
      <h1>
        你好
        <Button btnType='primary' onClick={() => close()}>close</Button> 
      </h1>
    );
  }
  return (
    <div>
      <div className="dialog-example-ct">
        <h1>例子1 使用Dialog原始组件</h1>
        <Button btnType='primary' onClick={() => setX(!x)}>click</Button> 
        <Dialog
          visible={x}
          buttons={
            [
              <Button btnType='primary' onClick={() => { setX(!x) }}>1</Button> ,
              <Button btnType='primary' onClick={() => { setX(!x) }}>2</Button> 
            ]
          }
          onClose={() => { setX(!x) }}
        >
          <div>dialog</div>
        </Dialog>
      </div>
      <div className="dialog-example-ct">
        <h1>例子2 使用Dialog原始组件（无默认遮罩层）</h1>
        <Button btnType='primary' onClick={() => setY(!y)}>click</Button> 
        <Dialog 
          visible={y} 
          closeOnClickMask={true} 
          buttons={
            [
              <Button btnType='primary' onClick={() => { setY(!y) }}>1</Button> ,
              <Button btnType='primary' onClick={() => { setY(!y) }}>2</Button> 
            ]
          }
          onClose={() => { setY(!y) }}
          enableMask={false}
        >
          <div>dialog</div>
        </Dialog>
      </div>
      <div className="dialog-example-ct">
        <h1>例子3 分别调用Dialog的动态组件alert，confirm，modal</h1>
        <Button btnType='primary' onClick={() => alert('1')}>alert</Button> 
        <Button btnType='primary' onClick={() => confirm('1', () => {
          console.log('yes');
        }, () => {
          console.log('no');
        })}>confirm</Button> 
        <Button btnType='primary' onClick={() => modal(<h1>你好</h1>)}>modal</Button> 
      </div>
      <div className="dialog-example-ct">
        <h1>例子4 在Dialog动态组件modal中调用Dialog API</h1>
        <Button btnType='primary' onClick={openModal}>modal</Button> 
      </div>
    </div>
  )
};