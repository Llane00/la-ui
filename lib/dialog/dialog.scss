@import '../helper';

.llane-ui-dialog {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  min-width: 20em;
  border-radius: 4px;
  background: white;
  z-index: 2;
  &-mask {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: fade-out(black, 0.5);
    z-index: 1;
  }
  &-header {
    font-size: 22px;
    padding: 8px 16px;
    border-bottom: 1px solid $main-color;
  }
  &-main {
    padding: 8px 16px;
    min-height: 6em;
  }
  &-footer {
    padding: 8px 16px;
    border-top: 1px solid $main-color;
    text-align: right;
  }
  &-close {
    position: absolute;
    bottom: 100%;
    left: 100%;
    transform: translate(-50%, 50%);
    display: flex;
    justify-content: center;
    align-items: center;
    width: 2em;
    height: 2em;
    color: #ffffff;
    background: $main-color;
    border-radius: 50%;
    .lui-icon {
      width: 2em;
      height: 2em;
    }
  }
}