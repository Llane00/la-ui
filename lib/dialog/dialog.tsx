import React, { Fragment, ReactElement } from 'react';
import ReactDOM from 'react-dom';
import './dialog.scss';
import Icon from '../icon/icon';
import { scopedClassMaker } from '../helpers/classes';

interface Props {
  visible: boolean;
  buttons?: Array<ReactElement>;
  onClose: React.MouseEventHandler;
  closeOnClickMask?: boolean;
  enableMask?: boolean;
}

const scopedClass = scopedClassMaker('llane-ui-dialog');
const sc = scopedClass;

const Dialog: React.FunctionComponent<Props> = (props) => {
  const onClickClose: React.MouseEventHandler = (e) => {
    props.onClose(e);
  };
  const onClickMask: React.MouseEventHandler = (e) => {
    if (props.closeOnClickMask) {
      props.onClose(e);
    }
  };
  const result = props.visible &&
    <Fragment>
      {props.enableMask && <div className={sc("mask")} onClick={onClickMask}></div>}
      <div className={sc("")}>
        <div className={sc("close")} onClick={onClickClose}>
          <Icon name="close" />
        </div>
        <header className={sc("header")}>
          提示
      </header>
        <main className={sc("main")}>
          {props.children}
        </main>
        {
          props.buttons && props.buttons.length > 0 &&
          <footer className={sc("footer")}>
            {props.buttons && props.buttons.map((button, index) => React.cloneElement(button, { key: index }))}
          </footer>
        }
      </div>
    </Fragment>;
  return (
    ReactDOM.createPortal(result, document.body) //传送门
  );
};
Dialog.defaultProps = {
  closeOnClickMask: false,
  enableMask: true
};
const modal = (content: any, buttons?: Array<ReactElement>, afterClose?: () => void) => {
  const onClose = () => {
    ReactDOM.render(React.cloneElement(component, [false]), div);
    ReactDOM.unmountComponentAtNode(div);
    div.remove();
  }
  const component = (
    <Dialog
      visible={true}
      buttons={buttons}
      onClose={() => {
        onClose();
        afterClose && afterClose();
      }}
      enableMask={true}
    >
      {content}
    </Dialog>);
  const div = document.createElement('div');
  document.body.append(div);
  ReactDOM.render(component, div);
  return onClose;
};
const alert = (content: string) => {
  const button = <button onClick={() => close()}>OK</button>;
  const close = modal(content, [button]);
};
const confirm = (content: string, yes?: () => void, no?: () => void) => {
  const onYes = () => {
    onClose();
    yes && yes();
  };
  const onNo = () => {
    onClose();
    no && no();
  };
  const buttons = [
    <button onClick={onYes}>yes</button>,
    <button onClick={onNo}>no</button>
  ];
  const onClose = modal(content, buttons, no);
};
export { alert, confirm, modal };
export default Dialog;