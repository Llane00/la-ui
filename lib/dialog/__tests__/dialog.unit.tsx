import renderer from 'react-test-renderer'
import React from 'react';
import { alert, confirm, modal } from '../dialog';
import Button from '../../button/button';

describe('dialog', () => {
  it('dialog alert组件快照测试', () => {
    const json = renderer.create(<Button btnType='primary' onClick={() => alert('1')}>alert</Button>).toJSON()
    expect(json).toMatchSnapshot()
  })
  it('dialog confirm组件快照测试', () => {
    const json = renderer.create(<Button btnType='primary' onClick={() => confirm('1', () => {
      console.log('yes');
    }, () => {
      console.log('no');
    })}>confirm</Button>).toJSON()
    expect(json).toMatchSnapshot()
  })
  it('dialog modal组件快照测试', () => {
    const json = renderer.create(<Button btnType='primary' onClick={() => modal(<h1>你好</h1>)}>modal</Button>).toJSON()
    expect(json).toMatchSnapshot()
  })

})
