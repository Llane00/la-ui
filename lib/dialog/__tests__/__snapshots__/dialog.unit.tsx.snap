// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dialog dialog alert组件快照测试 1`] = `
<button
  className="la-ui-button la-ui-button-primary"
  onClick={[Function]}
>
  alert
</button>
`;

exports[`dialog dialog confirm组件快照测试 1`] = `
<button
  className="la-ui-button la-ui-button-primary"
  onClick={[Function]}
>
  confirm
</button>
`;

exports[`dialog dialog modal组件快照测试 1`] = `
<button
  className="la-ui-button la-ui-button-primary"
  onClick={[Function]}
>
  modal
</button>
`;
